File: 7.reverse-integer.py
#
# @lc app=leetcode id=7 lang=python3
#
# [7] Reverse Integer
#

# @lc code=start
class Solution:
    def reverse(self, x: int) -> int:
        if abs(x) < 2147483647:
            if x//10 == 0 :
                return x
            else:
                init = 0
                nag = 0
                if x < 0:
                    nag = 1
                    x = -x
                while x > 0:
                    r = x%10
                    init = init * 10 + r
                    x = x//10
                if nag == 1:
                    init = -init
                if abs(init) < 2147483647:
                    return init
                return 0
        return 0
# @lc code=end

File: C#/7.reverse-integer.cs
/*
 * @lc app=leetcode id=7 lang=csharp
 *
 * [7] Reverse Integer
 */

// @lc code=start
public class Solution {
    public int Reverse(int x) {
        if(x == 0 || x == int.MinValue) return 0;
        int ans = 0;
        int pre_overflow = int.MaxValue / 10;
        int suf_overflow = int.MaxValue % 10;
        bool neg_tag = false;
        if(x < 0){
            neg_tag = true;
        }
        // return directly if x == 0
        x = Math.Abs(x);
        while(x > 0){
            int res = x % 10;
            // if the current answer > int.MaxValue % 10, in the following step, ans = ans * 10, it will cause overflow
            if(ans > pre_overflow) return 0;
            // if answer == int.MaxValue // 10, check the comming digit.
            if(ans == pre_overflow && res > suf_overflow) return 0;
            ans = ans * 10 + res;
            x = x / 10;
        }
        if(neg_tag == true){
            return -ans;
        }
        else return ans;
    }
}
// @lc code=end

